src/pint/toa.py

```diff
@@ -2072,34 +2063,33 @@ def apply_clock_corrections(
             )
         )
         corr = np.zeros(self.ntoas) * u.s
-        times = self.table["mjd"]
-        for ii, key in enumerate(self.table.groups.keys):
-            grp = self.table.groups[ii]
-            obs = self.table.groups.keys[ii]["obs"]
+        for obs, grp in self.get_obs_groups():
             site = get_observatory(
                 obs,
                 include_gps=include_gps,
                 include_bipm=include_bipm,
                 bipm_version=bipm_version,
             )
-            loind, hiind = self.table.groups.indices[ii : ii + 2]
             # First apply any TIME statements
-            for jj in range(loind, hiind):
+            for jj in grp:
```

Comment:
Is this loop necessary? can you do `self["mjd", grp] += time.TimeDelta(corr)`? (You might need both fancy and column indexing at once - `self["mjd", grp]`, or they might need to be separate - `self["mjd"][grp]`.)

Reply:
I had tried this and got various non-implemented errors when I tried it with the clock corrections.  Here I figured that a loop of some sort was needed anyway to deal with the str->float conversion.  That could be more compact but something similar would be needed.

Reply:
I looked at this again.  A few issues:
* I can't combine fancy & column indexing:
```
In [19]: t['mjd',:10]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-19-48bcf6d97e11> in <module>
----> 1 t['mjd',:10]

~/pythonpackages/PINT/src/pint/toa.py in __getitem__(self, index)
   1370                 return self.table[column]
   1371             else:
-> 1372                 return self.table[column, subset]
   1373         else:
   1374             r = []

~/opt/anaconda3/envs/pintdev3.6/lib/python3.6/site-packages/astropy/table/table.py in __getitem__(self, item)
   1669         else:
   1670             raise ValueError('Illegal type {} for table item access'
-> 1671                              .format(type(item)))
   1672 
   1673     def __setitem__(self, item, value):

ValueError: Illegal type <class 'tuple'> for table item access
```
* I can't add directly to the Column data:
```
In [25]: t['mjd'] + TimeDelta(10*u.ms*np.ones(len(t)))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-25-ee04eeefd5bf> in <module>
----> 1 t['mjd'] + TimeDelta(10*u.ms*np.ones(len(t)))

TypeError: unsupported operand type(s) for +: 'Column' and 'TimeDelta'
```
* I can't add to the `data` either:
```
In [31]: t['mjd'].data + TimeDelta(10*u.ms*np.ones(len(t)))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-31-5ae8f62b4d42> in <module>
----> 1 t['mjd'].data + TimeDelta(10*u.ms*np.ones(len(t)))

TypeError: unsupported operand type(s) for +: 'numpy.ndarray' and 'TimeDelta'
```
I haven't delved too far into this, so it could be user error.  But it also could be that our `Time` objects all have different locations so can't be a single `Time` array.

Reply:
- `:10` is slice indexing, not fancy, and the line error you're getting suggests that fancy indexing woudld work.
- Indeed Astropy `Column` objects don't support various things. 
- The last is nothing to do with any code we wrote, it's Astropy Time being fussy. You could convert the TimeDelta to days or make a Time object out of the Column.

Reply:
In fact I didn't bother to add slice indexing because, with TOAs in some unfriendly order, I expected users to be surprised if they tried to use slices. I don't think it would be very difficult.

Reply:
I did try the `TimeDelta` as just a quantity - still wouldn't work with the column.  Making a Time object out of the Column would (I think) lose the individual location for each entry.   In any case I'm not sure it's worth doing that and going back?